core/src/main/scala/org/apache/spark/rdd/CartesianRDD.scala

```diff
@@ -72,8 +74,21 @@ class CartesianRDD[T: ClassTag, U: ClassTag](
 
   override def compute(split: Partition, context: TaskContext): Iterator[(T, U)] = {
     val currSplit = split.asInstanceOf[CartesianPartition]
+
+    val key = RDDBlockId(rdd2.id, currSplit.s2.index)
+    val updatedBlocks = new ArrayBuffer[(BlockId, BlockStatus)]
+    def cachedValues(): Iterator[U] = {
+      SparkEnv.get.blockManager.memoryStore.unrollSafely(key,
+        rdd2.iterator(currSplit.s2, context), updatedBlocks) match {
+          case Left(arr) =>
+            arr.iterator.asInstanceOf[Iterator[U]]
+          case Right(it) =>
+            it.asInstanceOf[Iterator[U]]
+        }
+    }
+
     for (x <- rdd1.iterator(currSplit.s1, context);
-         y <- rdd2.iterator(currSplit.s2, context)) yield (x, y)
+         y <- new InterruptibleIterator(context, cachedValues())) yield (x, y)
```

Comment:
@squito Yes. I noticed that but have not committed updates yet. Interesting is, the performance gap before and after applying the remotely caching approach #5572, seems to be pulled closer in latest codebase. Because this issue is there for a while, I wonder if other improvement already makes this update not important on performance. I will test it again to see if this is correct. If so, this two PRs can be closed.

@tbertelsen Can you try latest Spark codebase for `RDD.cartesian` performance too?